### ISSUE TITLE
Bad interaction between RegExp input and setOptions

### ISSUE THREAD

**Issue by davisjam** [DmitrySoshnikov/regexp-tree#146]
```javascript
console.log('parse /a/');
regexpTree.parse(/a/);
console.log('set option, then parse /a/');
regexpTree.parser.setOptions({captureLocations: true}).parse(/a/);
```

Throws the following error (line 95 is the final line shown here):

```
parse /a/
set option, then parse /a/
Could not parse:
TypeError: string.lastIndexOf is not a function
    at Object.yyparse.onParseBegin (/home/jamie/Desktop/redos/analyze-regexps/regexp-features/node_modules/regexp-tree/src/parser/generated/regexp-tree.js:841:28)
    at Object.parse (/home/jamie/Desktop/redos/analyze-regexps/regexp-features/node_modules/regexp-tree/src/parser/generated/regexp-tree.js:672:13)
    at Object.<anonymous> (/home/jamie/Desktop/redos/analyze-regexps/regexp-features/report-regexp-features.js:95:58)
    at Module._compile (module.js:635:30)
    at Object.Module._extensions..js (module.js:646:10)
    at Module.load (module.js:554:32)
    at tryModuleLoad (module.js:497:12)
    at Function.Module._load (module.js:489:3)
    at Function.Module.runMain (module.js:676:10)
    at startup (bootstrap_node.js:187:16)
```

**Comment by davisjam** [comment]
Works fine if you do

```javascript
regexpTree.parser.setOptions({captureLocations: true});
regexpTree.parse(/a/);
```

Note that the example I'm using is modeled after this example from the README, which also fails at the moment.

```javascript
const regexpTree = require('regexp-tree');

const parsed = regexpTree
  .parser
  .setOptions({captureLocations: true})
  .parse(/a|b/);
```

I am using node v8.9.3.

**Comment by DmitrySoshnikov** [comment]
Yeah, good catch as well. It's because `regexpTree.parse` and `regexpTree.parser.parse` are two different functions.

The [regexpTree.parse](https://github.com/DmitrySoshnikov/regexp-tree/blob/master/index.js#L53) is a wrapper on top of the `regexpTree.parser.parse`, and does `toString` [coercion](https://github.com/DmitrySoshnikov/regexp-tree/blob/master/index.js#L54) before passing the argument to the `regexpTree.parser.parse` which always [expects](https://github.com/DmitrySoshnikov/syntax/blob/master/src/templates/lr.template.js#L132) the value as a string.

The `regexpTree.parser.parse` is an auto-generated code from the Syntax tool. I think it's safe to change it globally in the Syntax's templates, and do (in [lr.template.js](https://github.com/DmitrySoshnikov/syntax/blob/master/src/templates/lr.template.js#L132), and in [ll.template.js](https://github.com/DmitrySoshnikov/syntax/blob/master/src/templates/ll.template.js#L36) in Syntax repo):

```js
parse(string, parseOptions) {
  string = `${string}`;
  ...
```

In the [tokenizer.template.js](https://github.com/DmitrySoshnikov/syntax/blob/master/src/templates/tokenizer.template.js#L18):

```js
initString(string) {
  string = `${string}`;
  ...
}
```

This will fix the current issue with throw. I can take a look later, and will appreciate a PR of course :)

We probably should also rename the `setOptions` on the parser to `setGlobalOptions`, and probably even discourage using it (in favor of options per `parse` function call). It sets the global state between all `parse` calls, and can be a source of bugs. This change should be done in regexp-tree repo already.

**Comment by DmitrySoshnikov** [comment]
Actually, we can just patch the `parse` method in regexp-tree, [here](https://github.com/DmitrySoshnikov/regexp-tree/blob/master/src/parser/index.js#L11):

```js
const generatedParseFn = regexpTreeParser.parse.bind(regexpTreeParser);

/**
 * Parses a regular expression.
 *
 * Override original `regexpTreeParser.parse` to convert to string,
 * since in regexp-tree we may pass as a RegExp instance.
 */
regexpTreeParser.parse = function(regexp, options) {
  return generatedParseFn(`${regexp}`, options);
};
```

**Comment by DmitrySoshnikov** [comment]
Fixed in https://github.com/DmitrySoshnikov/regexp-tree/commit/2af8f847c9e58088daa502f77cbebae37c0b26a4, and in published in `v0.0.77`.
